### PR TITLE
Use the latest ReverseDiff

### PR DESCRIPTION
**Pull request by mohamed82008**
The latest version of ReverseDiff (https://github.com/JuliaRegistries/General/pull/18125) has a lot of the functionality in DistributionsAD. This PR removes the code already ported to RD and sets a lower bound on RD's version. Tests should pass after the RD PR is merged.

### CODE REVIEW

**Comment by mohamed82008** [comment]
There is an issue with the latest ReverseDiff version. I am on it.

**Comment by mohamed82008** [comment]
CC: @ChrisRackauckas

**Comment by ChrisRackauckas** [comment]
I'm also seeing something a bit odd: https://travis-ci.org/github/SciML/DiffEqFlux.jl/jobs/709405023

**Comment by mohamed82008** [comment]
Almost fixed it locally.

**Comment by mohamed82008** [comment]
Once tests pass locally, I will push the fixes.

**Comment by ChrisRackauckas** [comment]
No worries, thanks for being on it. Out of curiosity, what aspect does it seem to be pointing to?

**Comment by mohamed82008** [comment]
It was a name clash mostly. I defined `_add_to_deriv!` for the grad macro and another for broadcasting. Wrong methods were called. Also some functions were defined in propagation.jl and used in broadcast.jl but propagation.jl was included after broadcast.jl!

**Comment by mohamed82008** [comment]
The remaining failures are ambiguity errors when constructing `Wishart` and `InverseWishart` probably caused by the a change in Distribution.jl.

**Comment by mohamed82008** [comment]
I fixed all the errors locally. Once https://github.com/JuliaStats/Distributions.jl/pull/1148 gets merged and released, I will push the fixes. If the Distributions.jl maintainers don't like the PR, I will workaround it by defining `TuringMatrixBeta` but let's wait and see first.

**Comment by yebai** [comment]
We should rebase the changes in #95 before merging this PR.

**Comment by mohamed82008** [comment]
Fair enough.

File: src/matrixvariate.jl [hunk 1]
```diff
@@ -30,16 +30,20 @@ function TuringWishart(d::Wishart)
...
-    p = size(S, 1)
-    df > p - 1 || error("dpf should be greater than dim - 1.")
+function TuringWishart(df::Real, S::AbstractMatrix{<:Real})
     C = cholesky(S)
     return TuringWishart(df, C)
 end
+function TuringWishart(df::Real, S::AbstractPDMat{<:Real})
+    C = getchol(S)
+    return TuringWishart(df, C)
+end
```
Comment (devmotion) [inline src/matrixvariate.jl:43]: I'd suggest to just define
```julia
getchol(x::AbstractMatrix{<:Real}) = cholesky(x)
TuringWishart(df::Real, S::AbstractMatrix{<:Real}) = TuringWishart(df, getchol(S))
```
and remove the existing definitions of `TuringWishart(df::Real, S::AbstractMatrix{<:Real})` and `TuringWishart(df::Real, S::AbstractPDMat{<:Real})` completely.

File: src/matrixvariate.jl [hunk 2]
```diff
@@ -136,8 +140,8 @@ end
 
 ## Custom adjoint since Zygote can't differentiate through `@warn`
 # TODO: Remove when fixed upstream in Distributions
-ZygoteRules.@adjoint function Wishart(df::T, S::AbstractPDMat{T}, warn::Bool = true) where T
-    function _Wishart(df::T, S::AbstractPDMat{T}, warn::Bool = true) where T
+ZygoteRules.@adjoint function Wishart(df::T, S::AbstractPDMat{T}, warn::Bool = true) where {T <: Real}
+    function _Wishart(df::T, S::AbstractPDMat{T}, warn::Bool = true)
```
Comment (devmotion) [inline src/matrixvariate.jl:144]: Revert this change and add the inner `where` clause again? The function `_Wishart` was intended to be defined for all types `T`, possibly different from the outer type parameter `T`.

File: src/matrixvariate.jl [hunk 3]
```diff
@@ -175,10 +179,18 @@ getmatrix(p::PDMat) = p.mat
...
+function TuringInverseWishart(df::Real, Ψ::AbstractMatrix)
+    C = cholesky(Ψ)
+    return TuringInverseWishart(df, Ψ, C)
+end
+function TuringInverseWishart(df::Real, D::AbstractPDMat)
+    Ψ = getmatrix(D)
+    C = getchol(D)
+    return TuringInverseWishart(df, Ψ, C)
+end
+function TuringInverseWishart(df::T, Ψ::AbstractMatrix, C::Cholesky) where T<:Real
```
Comment (devmotion) [inline src/matrixvariate.jl:191]:
```suggestion
getmatrix(x::AbstractMatrix) = x

function TuringInverseWishart(df::T, x::AbstractMatrix) where T<:Real
    Ψ = getmatrix(x)
    C = getchol(x)
```
BTW it seems one could drop the `where` clause and just use `typeof(df)` in the function body?

File: src/matrixvariate.jl [hunk 4]
```diff
@@ -271,10 +283,18 @@ end
 
 Distributions.Wishart(df::TrackedReal, S::Matrix{<:Real}) = TuringWishart(df, S)
 Distributions.Wishart(df::TrackedReal, S::AbstractMatrix{<:Real}) = TuringWishart(df, S)
+Distributions.Wishart(df::Real, S::AbstractMatrix{<:TrackedReal}) = TuringWishart(df, S)
+Distributions.Wishart(df::TrackedReal, S::AbstractMatrix{<:TrackedReal}) = TuringWishart(df, S)
 Distributions.Wishart(df::Real, S::TrackedMatrix) = TuringWishart(df, S)
 Distributions.Wishart(df::TrackedReal, S::TrackedMatrix) = TuringWishart(df, S)
+Distributions.Wishart(df::Real, S::AbstractPDMat{<:TrackedReal}) = TuringWishart(df, S)
+Distributions.Wishart(df::TrackedReal, S::AbstractPDMat{<:TrackedReal}) = TuringWishart(df, S)
```
Comment (devmotion) [inline src/matrixvariate.jl:291]: This shouldn't be needed since `AbstractPDMat <: AbstractMatrix` since PDMats 0.10.
```suggestion
```

File: src/matrixvariate.jl [hunk 5]
```diff
@@ -271,10 +283,18 @@ end
...
 
 Distributions.InverseWishart(df::TrackedReal, S::Matrix{<:Real}) = TuringInverseWishart(df, S)
 Distributions.InverseWishart(df::TrackedReal, S::AbstractMatrix{<:Real}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::AbstractMatrix{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractMatrix{<:TrackedReal}) = TuringInverseWishart(df, S)
 Distributions.InverseWishart(df::Real, S::TrackedMatrix) = TuringInverseWishart(df, S)
 Distributions.InverseWishart(df::TrackedReal, S::TrackedMatrix) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::AbstractPDMat{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractPDMat{<:TrackedReal}) = TuringInverseWishart(df, S)
```
Comment (devmotion) [inline src/matrixvariate.jl:300]: Same here:
```suggestion
```

File: src/reversediff.jl [hunk 6]
```diff
@@ -249,15 +249,23 @@ function logpdf(d::MatrixBeta, X::AbstractArray{<:TrackedMatrix{<:Real}})
...
-InverseWishart(df::Real, S::TrackedMatrix) = TuringInverseWishart(df, S)
-InverseWishart(df::TrackedReal, S::TrackedMatrix) = TuringInverseWishart(df, S)
+Distributions.Wishart(df::TrackedReal, S::Matrix{<:Real}) = TuringWishart(df, S)
+Distributions.Wishart(df::TrackedReal, S::AbstractMatrix{<:Real}) = TuringWishart(df, S)
+Distributions.Wishart(df::Real, S::AbstractMatrix{<:TrackedReal}) = TuringWishart(df, S)
+Distributions.Wishart(df::TrackedReal, S::AbstractMatrix{<:TrackedReal}) = TuringWishart(df, S)
+Distributions.Wishart(df::Real, S::TrackedMatrix) = TuringWishart(df, S)
+Distributions.Wishart(df::TrackedReal, S::TrackedMatrix) = TuringWishart(df, S)
+Distributions.Wishart(df::Real, S::AbstractPDMat{<:TrackedReal}) = TuringWishart(df, S)
+Distributions.Wishart(df::TrackedReal, S::AbstractPDMat{<:TrackedReal}) = TuringWishart(df, S)
```
Comment (devmotion) [inline src/reversediff.jl:259]:
```suggestion
```

File: src/reversediff.jl [hunk 7]
```diff
@@ -249,15 +249,23 @@ function logpdf(d::MatrixBeta, X::AbstractArray{<:TrackedMatrix{<:Real}})
...
+
+Distributions.InverseWishart(df::TrackedReal, S::Matrix{<:Real}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractMatrix{<:Real}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::AbstractMatrix{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractMatrix{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::TrackedMatrix) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::TrackedMatrix) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::AbstractPDMat{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractPDMat{<:TrackedReal}) = TuringInverseWishart(df, S)
```
Comment (devmotion) [inline src/reversediff.jl:268]:
```suggestion
```

BTW are all other dispatches necessary?

File: src/matrixvariate.jl (hunk 1 above)
Comment (devmotion) [reply src/matrixvariate.jl:43]: Or maybe we could just use `cholesky` for all matrices? It is defined also for `AbstractPDMat`, so `getchol` could be removed completely and we just use
```julia
TuringWishart(df::Real, S::AbstractMatrix{<:Real}) = TuringWishart(df, cholesky(S))
```

File: src/matrixvariate.jl (hunk 3 above)
Comment (devmotion) [reply src/matrixvariate.jl:191]: Alternatively (see above), we could maybe use
```suggestion
getmatrix(x::AbstractMatrix) = x

function TuringInverseWishart(df::T, x::AbstractMatrix) where T<:Real
    Ψ = getmatrix(x)
    C = cholesky(x)
```

**Comment by devmotion** [comment]
I fixed some merge conflicts arising from recent changes in the master branch. Hopefully I didn't mess up anything :crossed_fingers:

**Comment by devmotion** [comment]
Anything else that you want to include in this PR @mohamed82008?

**Comment by mohamed82008** [comment]
LGTM, thanks @devmotion!
